MobileCenter/MobileCenterMac/Internals/Device/MSDeviceTracker.m

```diff
@@ -0,0 +1,316 @@
+    return [self device];
+  } else {
+
+    // This implements a binary search with complexity O(log n).
+    MSDeviceHistoryInfo *find = [[MSDeviceHistoryInfo alloc] initWithTOffset:toffset andDevice:nil];
+    NSUInteger index =
+        [self.deviceHistory indexOfObject:find
+                            inSortedRange:NSMakeRange(0, self.deviceHistory.count)
+                                  options:NSBinarySearchingFirstEqual | NSBinarySearchingInsertionIndex
+                          usingComparator:^(id a, id b) {
+                            return [((MSDeviceHistoryInfo *)a).tOffset compare:((MSDeviceHistoryInfo *)b).tOffset];
+                          }];
+
+    // All tOffsets are larger.
+    if (index == 0) {
+      return self.deviceHistory[0].device;
+    }
+
+    // All toffsets are smaller.
+    else if (index == self.deviceHistory.count) {
+      return [self.deviceHistory lastObject].device;
+    } else {
+      // Either the deviceHistory contains the exact toffset or we pick the smallest delta.
+      long long leftDifference = [toffset longLongValue] - [self.deviceHistory[index - 1].tOffset longLongValue];
+      long long rightDifference = [self.deviceHistory[index].tOffset longLongValue] - [toffset longLongValue];
+      if (leftDifference < rightDifference) {
+        --index;
+      }
+      return self.deviceHistory[index].device;
+    }
+  }
+}
+
+- (void)clearDevices {
+  @synchronized(self) {
+
+    // Clear persistence.
+    [MS_USER_DEFAULTS removeObjectForKey:kMSPastDevicesKey];
+
+    // Clear cache.
+    self.device = nil;
+    [self.deviceHistory removeAllObjects];
+  }
+}
+
+#pragma mark - Helpers
+
+- (NSString *)sdkName:(const char[])name {
+  return [NSString stringWithUTF8String:name];
+}
+
+- (NSString *)sdkVersion:(const char[])version {
+  return [NSString stringWithUTF8String:version];
+}
+
+- (NSString *)deviceModel {
+  size_t size;
+  sysctlbyname("hw.machine", NULL, &size, NULL, 0);
+  char *answer = (char *)malloc(size);
+  if (answer == NULL) {
+    return nil;
+  }
+  sysctlbyname("hw.machine", answer, &size, NULL, 0);
+  NSString *model = [NSString stringWithCString:answer encoding:NSUTF8StringEncoding];
+  free(answer);
+  return model;
+}
+
+- (NSString *)osName {
+  return @"macOS";
+}
+
+- (NSString *)osVersion {
+  NSString *osVersion = nil;
+
+#if __MAC_OS_X_VERSION_MAX_ALLOWED > 1090
+  if ([[NSProcessInfo processInfo] respondsToSelector:@selector(operatingSystemVersion)]) {
+    NSOperatingSystemVersion osSystemVersion = [[NSProcessInfo processInfo] operatingSystemVersion];
+    osVersion = [NSString stringWithFormat:@"%ld.%ld.%ld", (long)osSystemVersion.majorVersion,
+                                           (long)osSystemVersion.minorVersion, (long)osSystemVersion.patchVersion];
+  }
+#else
+  SInt32 major, minor, bugfix;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated"
```

Comment:
Maybe add a //TODO here. I think it's fine if use the HA stuff for now, let's set aside some time for that once the macOS SDK becomes a reality